src/Sparrow.Server/Allocators.Pool.NonBlocking.cs

```diff
@@ -0,0 +1,308 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Sparrow.Binary;
+using Sparrow.Threading;
+
+namespace Sparrow.Server
+{
+    /// <summary>
+    /// The NonBlockingPoolAllocator is a thread safe general pool allocator, which will hold all the memory it can during the process.
+    /// It will not keep track of allocations (except when running in validation mode), that means this allocator can leak if used improperly.
+    /// The idea is that we will store in concurrent stacks the blocks of memory and use non-blocking APIs to release and allocate from the pool.
+    /// </summary>
+    /// <typeparam name="TOptions">The options to use for the allocator.</typeparam>
+    /// <remarks>The Options object must be properly implemented to achieve performance improvements. (use constants as much as you can on configuration)</remarks>
+    public unsafe struct NonBlockingPoolAllocator<TOptions> : IAllocator<NonBlockingPoolAllocator<TOptions>, BlockPointer>, ILowMemoryHandler<NonBlockingPoolAllocator<TOptions>>, IRenewable<NonBlockingPoolAllocator<TOptions>>
+        where TOptions : struct, IPoolAllocatorOptions
+    {
+        private TOptions _options;        
+
+        // PERF: This should be devirtualized.        
```

Comment:
How, this is a interface field, so have to be boxed and have to go through the method dispatch

Reply:
Nope, this particular case will actually hit a special JIT code-path as long as you never return a TOptions based naked interface.

Reply:
This is `Allocator<TAllocator>`, which is a class, unless I missed something?
And the JIT don't know what the actual type will be here.

Reply:
The trick is that it knows when it gets resolved, which aint by the time the code is emitted for `Sparrow.Server` but when it is used like in https://github.com/ravendb/ravendb/pull/8921/files#diff-ae8fcae82d91040bb8935f9585e85fbfR87